Added details of client-side extension points with Forms 14. (#6140)
umbraco/UmbracoDocs: 14/umbraco-forms/developer/extending/adding-a-fieldtype.md

using Umbraco.Cms.Core.DependencyInjection;
using Umbraco.Forms.Core.Providers;

namespace MyProject;

public class Startup : IComposer
{
    public void Compose(IUmbracoBuilder builder)
    {
        builder.WithCollectionBuilder<FieldCollectionBuilder>()
            .Add<SliderFieldType>();
    }
}
```

> ## Partial view

Reviewer: ```suggestion
## Partial View
```

`IsMandatory` if set to `true` will provide client-side validation in the backoffice to ensure the value is completed.

When creating a field or other provider type, you might choose to inherit from an existing class. This could be if one of the types provided with Umbraco Forms almost meets your needs but you want to make some changes.

All setting properties for the Forms provider types are marked as `virtual`, so you can override them and change the setting values:

## Umbraco backoffice components

With Forms 14, aspects of the presentation and functionality of the custom field are handled by client-side components, registered via manifests:

- The preview, displayed on the form definition editor.
- The property editor UI used for editing the the submitted values via the backoffice.
- The property editor UI used for editing settings.
- A settings converter, that handles configuring the property editor and translating between the editor and persisted values.
- Translations for setting labels and descriptions.

> In order to create custom backoffice components for Umbraco 14, it's recommended to use a front-end build setup using Vite, TypeScript and Lit. Getting started with this is [described in the CMS documentation](https://docs.umbraco.com/umbraco-cms/tutorials/creating-your-first-extension#extension-with-vite-typescript-and-lit).

Reviewer: ```suggestion
To create custom backoffice components for Umbraco 14, it's recommended to use a front-end build setup using Vite, TypeScript, and Lit. For more information, see the [Creating your first extension](https://docs.umbraco.com/umbraco-cms/tutorials/creating-your-first-extension#extension-with-vite-typescript-and-lit) article.
```

> ### Field preview

Reviewer: ```suggestion
### Field Preview
```

const elementName = "my-property-editor-ui-number";

@customElement(elementName)
export class MyPropertyEditorUINumberElement
  extends UmbFormControlMixin<string>(UmbLitElement, undefined)
  implements UmbPropertyEditorUiElement
{
  private onChange(e: Event) {
    const newValue = (e.target as HTMLInputElement).value;
    if (newValue === this.value) return;
    this.value = newValue;
    this.dispatchEvent(new UmbPropertyValueChangeEvent());
  }

  render() {
    return html`<uui-input
      .value=${this.value ?? ""}
      type="number"
      @input=${this.onChange}
    ></uui-input>`;
  }
}

export default MyPropertyEditorUINumberElement;

declare global {
  interface HTMLElementTagNameMap {
    [elementName]: MyPropertyEditorUINumberElement;
  }
}
```

Again, it's registered via a manifest.

```javascript
const numberPropertyEditorManifest = {
    type: 'propertyEditorUi',
    alias: 'My.PropertyEditorUi.InputNumber',
    name: 'Number Input Property Editor UI',
    element: () => import('./property-editor-ui-number.element.js'),
    meta: {
        label: 'Number Input',
        icon: 'icon-autofill',
    },
};
export const manifests = [numberPropertyEditorManifest];
```

> ### Setting value editor

Reviewer: ```suggestion
### Setting Value Editor
```

{
  protected firstUpdated(
    _changedProperties: PropertyValueMap<any> | Map<PropertyKey, unknown>
  ): void {
    super.firstUpdated(_changedProperties);
    this.addFormControlElement(this.shadowRoot!.querySelector("input")!);
  }

  private onChange(e: Event) {
    const newValue = (e.target as HTMLInputElement).value;
    if (newValue === this.value) return;
    this.value = newValue;
    this.dispatchEvent(new UmbPropertyValueChangeEvent());
  }

  render() {
    return html`<input
      .value=${this.value ?? ""}
      type="color"
      @input=${this.onChange}
    />`;
  }
}

export default MyPropertyEditorUIColorElement;

declare global {
  interface HTMLElementTagNameMap {
    [elementName]: MyPropertyEditorUIColorElement;
  }
}
```

And once more, register via a manifest:

```javascript
const colorPropertyEditorManifest = {
    type: 'propertyEditorUi',
    alias: 'My.PropertyEditorUi.InputColor',
    name: 'Color Input Property Editor UI',
    element: () => import('./property-editor-ui-color.element.js'),
    meta: {
        label: 'Color Input',
        icon: 'icon-autofill',
    },
};

export const manifests = [colorPropertyEditorManifest];
```

> ### Setting value converter

Reviewer: ```suggestion
### Setting Value Converter
```

The following code shows the structure for these converter elements.

```javascript
import type { UmbPropertyValueData } from "@umbraco-cms/backoffice/property";

export class SliderSettingValueConverter {

  async getSettingValueForEditor(setting, alias: string, value: string) {
    return Promise.resolve(value);
  }

  async getSettingValueForPersistence(setting, valueData: UmbPropertyValueData) {
    return Promise.resolve(valueData.value);
  }

  async getSettingPropertyConfig(setting, alias: string, values: UmbPropertyValueData[]) {
    return Promise.resolve([]);
  }
}
```

It's registered as follows. The `propertyEditorUiAlias` matches with the property editor UI that requires the conversions.

```javascript
import { SliderSettingValueConverter } from "./slider-setting-value-converter.api";

const sliderValueConverterManifest = {
  type: "formsSettingValueConverter",
  alias: "My.SettingValueConverter.Slider",
  name: "Slider Value Converter",
  propertyEditorUiAlias: "My.PropertyEditorUi.Slider",
  api: SliderSettingValueConverter,
};

export const manifests = [sliderValueConverterManifest];
```

> ### Language files

Reviewer: ```suggestion
### Language Files
```

    sliderDefaultValueDescription: `Default value shown when the slider is displayed`,
    sliderHideStepValuesLabel: `Hide step values`,
    sliderHideStepValuesDescription: `Indicate whether the the field's label should be shown when rendering the form`,
    sliderBgColorLabel: `Background color`,
    sliderBgColorDescription: `Background color for the field`,
  },
}
```

Each different type of extension for Forms uses a different root value:

- Data sources - `formProviderDataSources`
- Export types - `formProviderExportTypes`
- Field types - `formProviderFieldTypes`
- Prevalue sources - `formProviderPrevalueSources`
- Recordset actions - `formRecordSetActions`
- Workflows - `formProviderWorkflows`

The language files are registered with:

```javascript
import type { ManifestLocalization } from "@umbraco-cms/backoffice/extension-registry";

const localizationManifests: Array<ManifestLocalization> = [
  {
    type: "localization",
    alias: "My.Localization.En_US",
    weight: -100,
    name: "English (US)",
    meta: {
      culture: "en-us",
    },
    js: () => import("./en-us.js"),
  },
];
export const manifests = [...localizationManifests];
```

> ### Registering the components

Reviewer: ```suggestion
### Registering the Components
```

> Finally you will need an entry point to your client-side components that will register the manifests with Umbraco's extension registry. For example:

Reviewer: ```suggestion
Finally, you will need an entry point to your client-side components that will register the manifests with Umbraco's extension registry. For example:
```

`SupportsPlaceholders` is a flag indicating whether the setting can contain ["magic string" placeholders](../magic-strings.md) and controls whether they are parsed on rendering.

`HtmlEncodeReplacedPlaceholderValues` takes effect only if `SupportsPlaceholders` is `true`. It controls whether the replaced placeholder values should be HTML encoded (as is necessary for rendering within content from a rich text editor).

`IsMandatory` if set to `true` will provide client-side validation in the backoffice to ensure the value is completed.

When creating a field or other provider type, you might choose to inherit from an existing class. This could be if one of the types provided with Umbraco Forms almost meets your needs but you want to make some changes.

All setting properties for the Forms provider types are marked as `virtual`, so you can override them and change the setting values:

> ## Umbraco backoffice components

Reviewer: ```suggestion
## Umbraco Backoffice Components
```

  getSettingValue(key: string) {
    return this.settings[key];
  }

  render() {
    return html`<div
      style="background-color:${this.getSettingValue("BgColor")}"
    >
      <uui-slider
        .min=${parseInt(this.getSettingValue("Min"))}
        .max=${parseInt(this.getSettingValue("Max"))}
        .step=${this.getSettingValue("Step")}
        .value=${this.getSettingValue("DefaultValue")}
        ?hide-step-values=${this.getSettingValue("HideStepValues") === "True"}
      ></uui-slider>
    </div>`;
  }

  static styles = css`
    div {
      padding: var(--uui-size-4);
    }
  `;
}

export default MyFieldPreviewSliderElement;

declare global {
  interface HTMLElementTagNameMap {
    [elementName]: MyFieldPreviewSliderElement;
  }
}
```

And it is registered via a manifest:

```javascript
import MyFieldPreviewSliderElement from './slider-preview.element.js';

const sliderPreviewManifest = {
    type: "formsFieldPreview",
    alias: "My.FieldPreview.Slider",
    name: "Forms UUI Slider Field Preview",
    api: MyFieldPreviewSliderElement,
    element: () => import('./slider-preview.element.js')
  };

  export const manifests = [sliderPreviewManifest];
```

> ### Field editor

Reviewer: ```suggestion
### Field Editor
```
